src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinActivityConversionExtensions.cs

```diff
@@ -211,64 +211,71 @@ internal struct TagEnumerationState : IActivityEnumerator<KeyValuePair<string, o
-                        return true;
+                        PeerServiceResolver.InspectTag(ref this, key, strVal);
+
+                        if (key == SpanAttributeConstants.StatusCodeKey)
+                        {
+                            this.StatusCode = StatusHelper.GetStatusCodeForTagValue(strVal);
+                            continue;
+                        }
+                        else if (key == SpanAttributeConstants.StatusDescriptionKey)
+                        {
+                            // Description is sent as `error` but only if StatusCode is Error. See: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk_exporters/zipkin.md#status
+                            this.StatusDescription = strVal;
+                            continue;
+                        }
+                        else if (key == ZipkinErrorFlagTagName)
+                        {
+                            // Ignore `error` tag if it exists, it will be added based on StatusCode + StatusDescription.
+                            continue;
+                        }
                     }
-                    else if (key == ZipkinErrorFlagTagName)
+                    else if (tag.Value is int intVal && tag.Key == SemanticConventions.AttributeNetPeerPort)
                     {
-                        // Ignore `error` tag if it exists, it will be added based on StatusCode + StatusDescription.
-                        return true;
+                        PeerServiceResolver.InspectTag(ref this, key, intVal);
                     }
-                }
-                else if (activityTag.Value is int intVal && activityTag.Key == SemanticConventions.AttributeNetPeerPort)
-                {
-                    PeerServiceResolver.InspectTag(ref this, key, intVal);
-                }
-
-                PooledList<KeyValuePair<string, object>>.Add(ref this.Tags, activityTag);
 
-                return true;
+                    PooledList<KeyValuePair<string, object>>.Add(ref this.Tags, tag);
+                }
             }
         }
 
-        private struct EventEnumerationState : IActivityEnumerator<ActivityEvent>
+        private struct EventEnumerationState
         {
             public bool Created;
 
             public PooledList<ZipkinAnnotation> Annotations;
 
-            public bool ForEach(ActivityEvent activityEvent)
+            public void EnumerateEvents(Activity activity)
             {
-                if (!this.Created)
+                var enumerator = activity.EnumerateEvents();
+
+                if (enumerator.MoveNext())
                 {
                     this.Annotations = PooledList<ZipkinAnnotation>.Create();
                     this.Created = true;
-                }
 
-                PooledList<ZipkinAnnotation>.Add(ref this.Annotations, new ZipkinAnnotation(activityEvent.Timestamp.ToEpochMicroseconds(), activityEvent.Name));
+                    do
```

Comment:
I love a good do/while... though maybe use `goto`s here instead?

Reply:
The funny thing is goto, while, do/while none of that really exists it is all jumps at the end of the day 🤣 PS: I did it this way to avoid having to check `this.Created` each time an item is processed in the inner loop.

Reply:
It's all just 1s and 0s at the end of the day... And even then 1s and 0s are a fiction. What's the meaning of life? How can we even be sure that `this.Created`?

Reply:
@alanwest I guess we would know it all when `this.Disposed` 😀

Reply:
We should start OpenTelemetry Quantum SIG.

Reply:
> We should start OpenTelemetry Quantum SIG.

I'm in, happy to sponsor it 😆

Reply:
> I'm in, happy to sponsor it 😆

Can you get Scott Bakula to co-sponsor?

![p7894024_b_h9_aa](https://user-images.githubusercontent.com/3676547/185466931-0ad64df6-b16b-4b4d-a434-1d7945558b66.jpeg)